torchdata/nodes/multi_dataset_weighted_sampler.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+from typing import Any, Dict, Iterator, Optional
+
+import torch
+from torchdata.nodes.base_node import BaseNode, T
+
+
+class MultiDatasetWeightedSampler(BaseNode[T]):
+    """A node that samples from multiple datasets with weights."""
+
+    _ARBITRARY_LARGE_NUMBER = 1000
+
+    def __init__(
+        self,
+        source_nodes: Dict[str, BaseNode[T]],
+        weights: Dict[str, float],
+        stopping_criterion: str = "CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED",
+    ) -> None:
+        self.source_nodes = source_nodes
+        self.weights = weights
+        self.stopping_criterion = stopping_criterion
+
+    def iterator(self, initial_state: Optional[Dict[str, Any]]) -> Iterator[T]:
+        self._it = self.Iter(self, initial_state)
+        return self._it
+
+    def get_state(self) -> Dict[str, Any]:
+        if self._it is None:
+            iter(self)
+        return self._it.get_state()
+
+    class Iter(Iterator[T]):
```

Comment:
In this pattern (i.e. defining an inner class which captures everything about the `Iterator` along with the parent `Node` class), what are the functionalities that the node class holds which are exclusive to it ?

Reply:
This might be dumb, but what is `BaseNode` was subclassing `Iterators` and not `Iterables`?

Having a user define state for both the node and then for the inner iterator _seems_ like it can open many scenarios of things getting hairy. Will lean on you to learn more